### DISCUSSION TITLE
Help using Pico and ENV IV SHT40

### DISCUSSION THREAD

**Discussion by wingsterz** [micropython/micropython#14334] [Using MicroPython]
Help please!  I am a newbie to Micropython and the world of Pis and struggling to get this temperature sensor working for a project.  Please help... kids are keen to get our home made incubator up and running and I have hit a wall.

The ENV IV sensor contains a SHT40 and a BMP280.  I'd be happy if I could just get the SHT40 working.  My wiring is straightforward, with male jumpers onto a simple breadboard to the Pico (no additional pull-up resistors).  

I2C scan reports the expected device addresses for both the SHT40 and BMP280.  But querying the SHT40 seems to make it fall over (line 10 is where it reads from sht.measurements).

Any suggestions for what to try please? 

Code:

import time
from machine import Pin, I2C
from micropython_sht4x import sht4x

i2c = I2C(1, sda=Pin(2), scl=Pin(3))  # Correct I2C pins for RP2040
sht = sht4x.SHT4X(i2c)
print("i2c scan results: ", i2c.scan())

while True:
    temperature, relative_humidity = sht.measurements
    print(f"Temperature: {temperature:.2f}°C")
    print(f"Relative Humidity: {relative_humidity:.2%}%")
    print("")
    time.sleep(0.5)

This code returns this when run:

i2c scan results:  [68, 118]
Traceback (most recent call last):
  File "<stdin>", line 10, in <module>
  File "/lib/micropython_sht4x/sht4x.py", line 157, in measurements
OSError: [Errno 5] EIO

**Comment by robert-hh** [comment]
EOI is usually a communication error and most of the time by missing pull-up resistors at the SDA and SCL lines.
Since the scan returns proper numbers, the wiring is not completely wrong. Then, you could check if your SHT40 driver uses the address returned by scan().  For the SHT4x, 68 is the default address, 69 the alternative. For the BME280 it's 118.

**Reply by wingsterz** [reply]
Thank you Robert, very good of you to reply!  The SHT40 driver is reporting the right address - 68, 118 as expected so as you say the wiring is not completely wrong.  I don't have extra pull-up resistors, and I can't say I fully understand them but the PICO is all 3.3v not 5v, and the SHT40 I have is on a commercial enclosure (ENV IV) which I would hope has the pull-ups on them if needed.  Will contact Pi Hut to see if they can confirm.

**Reply by robert-hh** [reply]
The schematics and pictures of the ENV IV show pull up resistors, and the documentation confirms that the I2C address is 68 (0x44). Which library do you use for the SHT4x? How long is the wire to the ENV IV module?

Unless the wire is really long, there should not be a basic communication problem. But you can make a test by reducing the freq of I2C to a really low value, e.g. with `i2c = I2C(1, sda=Pin(2), scl=Pin(3), freq=1000)`. The default is 400_000. If that works, then there is a wiring problem.

**Reply by robert-hh** [reply]
I looked into the SHT4x data sheet to confirm, that 400kHz is supported, and it is. I could not tell whether there is a problem with RP2 and clock stretching. If it is, SoftI2C should work. Just try `i2c = SoftI2C(sda=Pin(2), scl=Pin(3))`.

**Reply by wingsterz** [reply]
Thank you Robert

I tried the low frequency for I2c, no impact.

SoftI2C returned the I2C scan address well, but fell over on querying the sensor, this time with a slightly different ERRNO of 19 ENODEV

Next I think I'll try replacing the wires as a last ditch as beetle gig suggests

**Comment by beetlegigg** [comment]
A couple of weeks ago I got this sensor - ENV IV Unit with Temperature Humidity Air Pressure Sensor (SHT40+BMP280).
I've only used the SHT40 thus far and I used a micropython library found here: 
https://github.com/jposada202020/MicroPython_SHT4X/tree/master/micropython_sht4x
The code test to get the temperature I used was:
```
#import machine
from machine import Pin, I2C
import sht4x

# i2c pins on picowbel stemma qt / qwiic
sda_pin = Pin(4) # grove white
scl_pin = Pin(5) # grove yellow

sht4x_address = 0x44
i2c = I2C(0,scl=scl_pin, sda=sda_pin)

sensor = sht4x.SHT4X(i2c)

def read_sht4x():
    temp, humi = sensor.measurements
    return(round(temp,1), round(humi,1))

if __name__ == '__main__':
    import time
    while True:
        temp, hum = read_sht4x()
        print('Temperature: ',temp)
        print('Humidity:    ',hum)
        time.sleep(10)

```
I hope this helps.
Edit - I see I did not actually use the sht4x_address when initialising the sensor, but the default address in the library I refer to is 0x44. (68 decimal)
Edit again :) - I should mention that my unit came installed on a board wrapped in a plastic enclosure with a grove connector.  The board would have the necessary pull up resistors already connected.  This was the board: 
https://thepihut.com/products/env-iv-unit-with-temperature-humidity-air-pressure-sensor-sht40-bmp280

**Reply by wingsterz** [reply]
Thank you Beetlegig, I too am using the ENV IV board, though directly to a PICO H, and without the Adafruit 'cowbell' that you seem to be using.  I had to modify the PIN numbers in your code to get the i2c scan working, but ultimately it still fell over with the ERRNO 5 sadly!  I do see what I expect are pull-up resistors in the official ENV IV schematic, they seem to be 4.7kohm resistors on SDA and SCL.  In a world where most things work out of the box this is quite a different experience!  Glad yours is working :)

#import machine
from machine import Pin, I2C
from micropython_sht4x import sht4x
import time

# i2c pins on picowbel stemma qt / qwiic
sda_pin = Pin(2) # grove white
scl_pin = Pin(3) # grove yellow

sht4x_address = 0x44
i2c = I2C(1,scl=scl_pin, sda=sda_pin)

time.sleep(1)
print(i2c.scan())

sensor = sht4x.SHT4X(i2c)

def read_sht4x():
    temp, humi = sensor.measurements
    return(round(temp,1), round(humi,1))

if __name__ == '__main__':
    while True:
        temp, hum = read_sht4x()
        print('Temperature: ',temp)
        print('Humidity:    ',hum)
        time.sleep(10)


Returns this:-
[68, 118]
Traceback (most recent call last):
  File "<stdin>", line 24, in <module>
  File "<stdin>", line 19, in read_sht4x
  File "/lib/micropython_sht4x/sht4x.py", line 157, in measurements
OSError: [Errno 5] EIO

**Reply by beetlegigg** [reply]
The cowbell is just a breakout board that mounts to the pico where one can mount components and it has an included i2c port wired in to one set of the pico's i2c pins.  I used a grove to qt i2c connector to connect up the sensor to that arrangement.    I do have another of these sensors and just connected it  directly to another pico via a connector that had grove to dupont connectors.  This worked ok and  I've not experienced any issues with my two sensors,  but maybe you have a duff one.  If you have not done so,  It might be worth you swapping out all your connection cables  for others  to eliminate something amiss with your wiring.  Also check that the library you are using the same as the one I linked to, I expect they are but worth a check.

**Reply by wingsterz** [reply]
Appreciate your extra responses and checks beetlegigg!  I too am using the jposada202020 code, which is the package found under PyPI for SHT40.  Great idea to replace wiring .... but it didn't help.  Will go back to the vendor and see if a replacement device has better luck.

**Reply by beetlegigg** [reply]
Just a thought, and probably you are fully aware of the comment I make, but as it popped into my mind I may as well add it to the conversation :).   I power my sensor from the picos 3v3 out and it works ok.  But the spec indicates powering it with 5v. The schematic shows an on board voltage regulator and shows the i2c output is 3v3 so ok for the pico pins.  Maybe you are also powering it via 3v3 but, for whatever reason, your unit could do with being powered by 5v ??   And of course a good ground connection is essential twixt the pico and sensor.  Also don't trust any breadboard connected wires, they quite often go wobbly.

**Comment by jboiron** [comment]
I had the same problem than you (with a breakout board) and after reading the Sentiron documentation, i had the following line : time.sleep(0.01) at line 153 of the sht4x micropython driver from jposada202020 and it works very well (it is the recommended value by Sentiron)

**Reply by rogerw-gh** [reply]
The PyPI version of this code is lacking the additional time.sleep(0.1) that the now deprecated Github repo contains.  I found the problem, eventually, having installed the PyPI version, and got exactly the same error reported by @wingsterz , then noticed the additional time.sleep, manually editing sht4x.py to add the extra line, then copying the new file to the Pico [in my case a Pimoroni Badger].  I used rshell to manually build the code which returned numbers, not errors.

**Comment by TimothyHowlett** [comment]
This, a cut down version of jposada202020's code, works fine for basic temperature and humidity readings:
```py
import time
import struct
from machine import Pin, I2C

i2c = I2C(1, sda=Pin(26), scl=Pin(27))

class SHT4X:
    def __init__(self, i2c, address: int = 68) -> None: # default address for SHT40 = 68 = 0x44
        self._i2c = i2c
        self._address = address
        self._data = bytearray(6)

    @property
    def measurements(self) -> tuple[float, float]:
        self._i2c.writeto(self._address, bytes([0xFD]), False) # command = 0xFD
        time.sleep(0.2)
        self._i2c.readfrom_into(self._address, self._data)
        temperature, temp_crc, humidity, humidity_crc = struct.unpack_from(">HBHB", self._data)
        temperature = -45.0 + 175.0 * temperature / 65535.0
        humidity = -6.0 + 125.0 * humidity / 65535.0
        humidity = max(min(humidity, 100), 0)
        return temperature, humidity

    def reset(self): # Reset the sensor
        self._i2c.writeto(self._address, bytes([0x94]), False) # _RESET = 0x94
        time.sleep(0.1)

sht = SHT4X(i2c)
sht.reset()

while True:
    temperature, relative_humidity = sht.measurements
    print(f"Temperature: {temperature:.1f}°C")
    print(f"Relative Humidity: {relative_humidity:.1f}%")
    time.sleep(2)
```
